### PR TITLE
OTP challenge support

### PR DESCRIPTION
**Pull request by hryamzik**
This PR is based on @llange [code](https://github.com/llange/Tunnelblick/tree/handle_challenge_response_protocols).
I just added support for using secret and storing it in keychain.

Addresses #267, #417, #300.

Also eventually fixed log message for passwordKeychain save failure.

### CODE REVIEW

**Comment by jkbullard** [comment]
@hryamzik - Thank you very much! I will look at this as soon as I can but am in the middle of several other things right now, so it will probably be at least a few days before I get a chance.

**Comment by hryamzik** [comment]
Good, thanks!

@llange, it would be awesome if you could take a look too, especially at this [code](https://github.com/llange/Tunnelblick/blob/handle_challenge_response_protocols/tunnelblick/VPNConnection.m#L2644) – I guess `echoResponse` parameter could be skipped at all.

**Comment by johnkeates** [comment]
This looks really nice. I wish I had more time to support or contribute to Tunnelblick to do this kind of PR.

**Comment by jkbullard** [comment]
@hryamzik - Thanks again. I've take a quick look at it and have some questions/comments:

Can you explain why the "Aerogear" is used for? What does it do that llange's code doesn't? What OTPs does this support? Does it support hardware OTPs and software OTPs? In other words, a bit more about the purpose of the code and why it does it the way it does.

In general I don't like to add additional dependencies to Tunnelblick. We already have many that are maintained separately or incorporated into the Tunnelblick source code. But because Aerogear is needed for this, I'm OK with adding it.

However, I do not like the direct commits of "modified" versions of Aerogear (assuming I understand this correctly). I would want a commit the original, unmodied Aerogear code with a reference to where it came from, and then a separate commit (or maybe even a commit per-change or per-file if there are many different reasons for the changes) to make modifications to it for Tunnelblick. That way we can track changes in the original code and more easily make corresponding changes to the modified-for-Tunnelblick code.

As a general rule, I prefer more granular commits. For example I would prefer "Cleaned up, added aerogear library" as two separate commits. (I admit I don't always do that myself, but in most cases that's due to "squashing" I've done on commits on my personal repositories.) The "Implemented OTP login with secret" is OK, though.

And thanks for the "Fixed log message for passwordKeychain save failure" commit. (The dangers of copy/paste). I'd be happy to accept a PR for that separately right now.

All this without actually looking at the code yet! I hope to do that in the next few days and will respond here when I've done that.

Thanks again! I hope this didn't sound discouraging; I do like your submission!

**Comment by hryamzik** [comment]
Hey @jkbullard, thanks for looking at this!

>Can you explain why the "Aerogear" is used for?
>However, I do not like the direct commits of "modified" versions of Aerogear

This code is not modified and copied as is. The idea is to support TOTP generation, that's how google authenticator works. So this code gives an option to save the secret and let Tunnelblick generate one time passwords itself when they are requested. I had an idea to hide this option with some extra keys (like option key pressed) but that was too complicated with an NSAlert.

Do you want me to rework commit history? That may be complicated due to large xml settings files changes.

**Comment by jkbullard** [comment]
> This code is not modified and copied as is.

OK, but where was it copied from? If it came from the official Aerogear [site](https://aerogear.org/) (I could only find versions for iOS, not macOS, and written in Swift, not Objective-C), why does it include lines like:

`//  Taken with small modifications`

(and where was _it_ taken from, and who made those modifications, and what were they?)

Don't do anything with the commit history until I look at the actual code, though.

Thanks again!

**Comment by hryamzik** [comment]
Here's the [link](https://github.com/aerogear/aerogear-ios-otp). Library is also available as POD however it doesn't recognise osx as a valid target.

**Comment by snowrider311** [comment]
Hi, 

First of all, thanks for contributing this - it's very helpful and I'm glad you undertook the work!

My use case is slightly different from yours, I guess - I **do** want support for challenge/response in order to collect a TOTP token, but I specifically **don't** want our users to be able to save TOTP secrets (in my view that defeats the purpose of requiring a TOTP token). I added a couple of comments on the code related to this.

I think with a few minor changes, this PR could address both use cases pretty well.

Thanks!

**Comment by troyready** [comment]
I think a case could be made for not including the secret saving functionality at all - many (most?) security policies would prohibit it.

File: tunnelblick/VPNConnection.m [hunk 1]
```diff
@@ -2631,6 +2633,46 @@ -(void) indicateWeAreHookedUp {
...
+    NSAlert *alert = [NSAlert alertWithMessageText: challenge
+                                     defaultButton:@"OK"
+                                   alternateButton:@"Cancel"
+                                       otherButton:nil
+                         informativeTextWithFormat:@""];
+    [alert setShowsSuppressionButton:YES];
+    [[alert suppressionButton] setTitle:NSLocalizedString(@"Input secret and save it in keychain","Challenge secret checkbox")];
+    
+    NSTextField *input;
+    input = [[NSSecureTextField alloc] initWithFrame:NSMakeRect(0, 0, 235, 24)];
```
[the remark this answers is not in the document]
Comment (hryamzik) [reply tunnelblick/VPNConnection.m]: That's tricky with current implementation but generally it's possible.

**Comment by hryamzik** [comment]
I would give user a choice. Generally that's less secure, however native openvpn client doesn't have keychain support and that may lead to passwords or certificates stored in plain text. OTP may be implemented to protect against that cases primarily.

You can also use policies to clearly restrict saving OTP secrets and QR codes on users devices except mobile phone apps.

I think that using a secret instead of an OTP should be hidden behind an extra hot key press. It turned out to be more tricky to implement with NSAlert than I expected so for now I had to give up.

P.S. don't forget that even 1passord provides OTP support. And that rocks.

**Comment by johnkeates** [comment]
I think giving users a choice would be the best way too. But if this needs to be pre-configured policy-wise, would it be an idea to have some sort of config option that allows/disallows this choice? I know users would be able to simply edit the file, but at least this would be the user breaking the policy and not the client.

**Comment by xvolte** [comment]
Hello,

i'm interested to have the GoogleAuth working.
My configuration on windows works correctly with the official openvpn client.

Can you give instruction on how to install / compile your modified version ?

is there a way to have a precompiled dmg with your fixes ?
or, should i compile from the source and replace some of the files by yours ? (and if so, how to do it?)

thanks in advance,

**Comment by hryamzik** [comment]
[Tunnelblick.dmg.zip](https://github.com/Tunnelblick/Tunnelblick/files/2026144/Tunnelblick.dmg.zip)

@xvolte give it a try.

**Comment by midN** [comment]
Why is this still not merged? :)

**Comment by jkbullard** [comment]
It has not been merged because:
 1. It is based on code (the "@llange [code](https://github.com/llange/Tunnelblick/tree/handle_challenge_response_protocols)") that has not been merged.
 1. It isn't clear to me where the supposed Aerogear code came from (Aerogear is iOS, not macOS and is written in Swift, not Objective-C), and what modifications were made to that code.
 1. I am skeptical about saving TOTP in the Keychain. I think it should not be allowed, or disallowed by default and be an "admin-only" preference that can only be changed by a computer admin.

**Comment by johnkeates** [comment]
I agree with 3, don't allow normal users to save the TOTP secret. It completely defeats the purpose of MFA.

1. Should be solvable

2. Needs real work I suppose, I bet someone with Swift + Obj-C/Cocoa knowledge would have to hammer on this and provide sources before this comes close to merging.

**Comment by faucct** [comment]
I think that separating OTP challenge support feature from the rest would reduce the amount of changes and let this be merged sooner.

**Comment by jkbullard** [comment]
> I think that separating OTP challenge support feature from the rest would reduce the amount of changes and let this be merged sooner.

Yes.

**Comment by hryamzik** [comment]
@pmoust 1password stores OTP secrets as well. I believe this option should be hidden and shown with some keys pressed, but that wasn't strait-forward to implement in an alert. For now I don't have time to improve this UI. However I don't want to totally loose ability to store secret.

As soon as native client doesn't support keychain at all people sometimes use simple passwords (and reuse them) or store them insecurely. OTP is helpful in this case. Tunnelblick + keychain + OTP secret in it doesn't look worse for me than simple/reused/bad-stored password + regular OTP.

Any UI improvements and PR splits (OTP support + proper UI to store secret) are welcomed.

**Comment by johnkeates** [comment]
OTP isn't as much a client thing as a server/service provider thing. Also, 1Password is a credential provider so it makes sense for it to have the secrets (so you can use 1Password as your code generator), but Tunnelblick needs to authenticate which means that having one or more factors is needed to ensure any single breach doesn't leak all credentials needed to authenticate.

Currently, there are multiple layers:
- TLS key for packet authentication
- Certificates for user authentication
- Username + Password voor person authentication via knowledge
- OTP for person authentication via possession 

Often, the TLS key and certs are stored on the filesystem unencrypted and not password-protected, so while a good first barrier for clients that try to login using nothing but username+password, it doesn't help against local attacks. For the username+password, it often is stored locally in the keychain, with the idea to use it as a password manager which is locked by the user of the system. Basically not more or less safe than using the credentials directly (because you still only need to compromise a single set of creds - the user/pass combo of the host system). This leaves OTP tokens as the only off-system credential that cannot be compromised on the system itself, which is the whole idea of an OTP token.

Allowing an OTP token or its secret to be stored or generated automatically completely defeats the purpose and you might as well stop using it. If you are using it for work, your security department probably makes the rules, and I bet the rules say you are not supposed to bypass the concept of multifactor authentication by having it on the system and automatically enter it on-demand.

**Comment by pmoust** [comment]
I echo @johnkeates stance (esp. the last bits).

FWIW I don't agree with 1Password's model, it can essentially violate MFA. All security policies on 2FA can be violated with `expect` , `oathtool --totp -b` etc people do that in the wild far more frequently that one would _expect_ (pun intended). I don't see why that's an argument countering my points.

**Comment by hryamzik** [comment]
@johnkeates 
> Allowing an OTP token or its secret to be stored or generated automatically completely defeats the purpose and you might as well stop using it.

I guess the reasonable option here is to set keychain policy to "always ask for permission".

And I admit that secret option shouldn't be showed OOTB, only with some special keys pressed (?) and may be in admin mode only.

However I didn't manage to get required controls on NSAlert in a reasonable amount of time, that's much more difficult than tuning iOS UIViews.

@pmoust the question is how secure are these approaches. I've implemented something similar once for [ssh over centrify](/hryamzik/sshit) and did store secret in keychain but I see lot's of people not taking care of that step. So they end up with unencrypted unprotected secrets, that looks much worse for me than 2fa secret in a keychain.

Let's do the following:
1. Feel free to fork my branch and create a new PR containing just 2fa support
2. I would really appreciate mentioned UI improvements for making storing 2fa secret not that easy to find

**Comment by gbdlin** [comment]
I think good solution will be to not store TOTP secret in keychain at all, but allow somehow TOTP code to be provided by some external script. That will also enable easy integration of hardware OTP and won't violate policy of OTP by default (additionally we can add annotation that script usage is designed for hardware OTP, but won't restrict it to that to make everyone happy).

That will still leave an option for someone in real need to violate 2FA and provide script to generate TOTP on actual machine (secret in script can be still fetched from keychain)

**Comment by hryamzik** [comment]
The problem here is that I'm not aware of any useful storages other than a keychain. 1password doesn't give us an API for some reason. So that external script functionality looks just like an extra option for me that may be useful for hardware tokens and other hacks making this brilliant tool even more flexible.

**Comment by gbdlin** [comment]
You can still use keychain in script. Also that will improve security because tunnelblick will never be aware of secret and there won't be possibility to one security breach to leak all of your credentials to VPN.

There are CLI and APIs for other password managers, like LastPass, so there is something to gain by that approach.

If that's the solution that will be approved and merged soon, maybe it is worth trying. There is not much to loose, only some checkbox(es) to save your secret in keychain automatically and much more to gain (possibility to improve secutity and get support for hardware tokens).

**Comment by johnkeates** [comment]
Stop trying to store the TOTP stuff, problem solved. All that is needed is:

1. When the challenge request is received, pop up a text box and a Ok + Cancel buttom
2. When the user clicks Ok, send whatever the user filled into the text box to the OpenVPN Server

**Comment by xvolte** [comment]
Agree with @johnkeates

**Comment by gbdlin** [comment]
I agree that this solution is sufficient for TOTP, but it won't be usable for any hardware-based 2FA (yubikeys etc). That's the main reason why I'm opting in for additional script-based solution.

**Comment by jkbullard** [comment]
I've closed this PR for several reasons:
 1. Static challenge/response has been added via PR #464, which was also based on the @llange code.
 2. Dynamic challenge/response has been added via 2d9f0e4159c66288600a468c468f9729869d025a.
 3. No provenance for the Aerogel code has been submitted.
 4. Saving the OTP in the Keychain is, in my view, a bad idea. It certainly should not be available as a default.

**Comment by hryamzik** [comment]
> No provenance for the Aerogel code has been submitted.

That's not true BTW.

**Comment by jkbullard** [comment]
@hryamzik - The [link](https://github.com/aerogear/aerogear-ios-otp) previously provided is for iOS.

**Comment by hryamzik** [comment]
I actually did use this code, it's compatible as doesn't have any ios framework specific components.

**Comment by johnkeates** [comment]
Points 1 and 2 make this issue redundant I guess, cool.

**Pull request closed without merging**